usr/lib/api/api_interface.c

```diff
@@ -56,22 +56,37 @@ void api_init();
 
 API_Proc_Struct_t *Anchor = NULL;       // Initialized to NULL
 unsigned int Initialized = 0;   // Initialized flag
-pthread_mutex_t GlobMutex;      // Global Mutex
+pthread_mutex_t GlobMutex = PTHREAD_MUTEX_INITIALIZER; // Global Mutex
 CK_FUNCTION_LIST FuncList;
 
 int slot_loaded[NUMBER_SLOTS_MANAGED];  // Array of flags to indicate
                                        // if the STDLL loaded
 
-// For linux only at this time... if it works out we can get rid
-// of the stupid pid tracking.... Linux we kind of have to do this
-// since new threads are processes also, and we will be hosed
+CK_BBOOL in_child_fork_initializer = FALSE;
+
 void child_fork_initializer()
 {
-    if (Anchor) {
-        free(Anchor);
-        Anchor = NULL;
-    }
-}
+    /*
+     * Reinitialize trace so that the trace output appears under the new
+     * process's trace file, and not in the parent process's once.
+     * C_Finalize will pass the new trace handle to the tokens, so that their
+     * finalization code will also trace into the new trace file.
+     */
+    trace_finalize();
+    trace_initialize();
+    /*
+     * Terminate all slots. This will also free the Anchor.
+     * A forked child needs to call C_Initialize again to use Opencryptoki.
+     * Indicate that we are in the fork initializer.
+     * A forked child process has the same mapped memory as the parent, but the
+     * reference count as part of the shared memory was not increased during
+     * the fork, so we should not decrement it when terminating.
```

Comment:
I personally found the explaination of what we are not doing confusing. Id rather say that the child is no PKCS11 application after fork (so no refcount needed). If it becomes a PKCS11 application by calling Initialize, refcount will be handled just like in the parent, right ?

Reply:
rephrased the comment, please take a look